Enhance Performance Evaluation Module

Yosri-Ben-Halima opened the pull request:
Changes involve:
- Refactor into PerformanceMetrics class for modular design
- Add benchmark metrics calculation
- Add report() method for strategy vs benchmark comparison table
- Add average drawdown and drawdown duration metrics
- Add return distribution stats: skewness, kurtosis, VaR, CVaR
- Add Calmar ratio calculation
- Replace alpha/beta with linear regression method
- Add rolling alpha/beta for dynamic risk analysis
- Standardize units and formatting for returns, drawdowns, and ratios

Commit:
Enhance Performance Evaluation Module:
- Refactor into PerformanceMetrics class for modular design
- Add benchmark metrics calculation
- Add report() method for strategy vs benchmark comparison table
- Add average drawdown and drawdown duration metrics
- Add return distribution stats: skewness, kurtosis, VaR, CVaR
- Add Calmar ratio calculation
- Replace alpha/beta with linear regression method
- Add rolling alpha/beta for dynamic risk analysis
- Standardize units and formatting for returns, drawdowns, and ratios

Commit:
Add test for max_drawdown with proper NaN handling

- Implements a unit test for `PerformanceMetrics.max_drawdown()`
- Includes sanity check for non-negative drawdown
- Handles potential NaN comparisons correctly using `math.isnan`

Commit:
Update requirements.txt

konradbachusz commented:
Thank you for your contribution, @Yosri-Ben-Halima.

I try to implement automatic unit tests for every new Python function that I write. To ensure stability of the solution, those unit tests fail. with each commit.

Currently it fails because one of the packages that you import is not in requirements.txt. https://github.com/konradbachusz/algorithmic-trading-utilities/actions/runs/17517522511/job/49756961658 

Could you please update requirements.txt with a specific version?

konradbachusz commented on algorithmic_trading_utilities/common/portfolio_ops.py:
```diff
@@ -8,284 +8,295 @@
     from data.yfinance_ops import get_sp500_prices
 except ImportError:
-    from algorithmic_trading_utilities.data.yfinance_ops import get_sp500_prices
+    from algorithmic_trading_utilities.data.yfinance_ops import get_sp500_prices  # noqa: F401
```
We don't seem to be using get_sp500_prices. What are we using as a benchmark in this branch?

Yosri-Ben-Halima replied:
I moved the benchmark data out so it’s passed explicitly as a parameter rather than being automatically set to the S&P 500. This makes the module more modular and avoids hardcoding assumptions. That’s why get_sp500_prices isn’t being used in this branch.

konradbachusz commented on algorithmic_trading_utilities/common/portfolio_ops.py:
```diff
@@ -8,284 +8,295 @@
+        portfolio_equity : np.ndarray
+            Daily portfolio equity values.
+        benchmark_equity : np.ndarray, optional
+            Daily benchmark equity values (for alpha/beta), default None.
```
Can we please default to S&P 500 that you can get from get_sp500_prices?

Yosri-Ben-Halima replied:
I’ll reintroduce the S&P 500 as a default benchmark value in case none is specified, so users still get an out-of-the-box experience while keeping the flexibility to pass their own benchmark data.

konradbachusz commented on algorithmic_trading_utilities/common/portfolio_ops.py:
```diff
@@ -8,284 +8,295 @@
+        )
+
+    @staticmethod
+    def _daily_returns(equity: np.ndarray) -> np.ndarray:
```
Can we please include Google-style docstrings for ALL functions in this branch? You can follow the style from any other functions in this repo and use GenAI to write the docstrings for you.

Yosri-Ben-Halima replied:
Okay sure, will take care of that asap

konradbachusz replied:
Can we please use Google-style docstrings for all of the remaining functions?

Example
```
        """
        Compute daily returns from equity curve.

        Args:
            equity (pd.Series): Series of equity values.

        Returns:
            pd.Series: Series of daily returns.
        """
```

You can use GenAI to do this for you

Yosri-Ben-Halima replied:
Done

konradbachusz commented on algorithmic_trading_utilities/common/portfolio_ops.py:
```diff
@@ -8,284 +8,295 @@
+            return np.array([])
+        return np.diff(equity) / equity[:-1]
+
+    # region - Basic Metrics & Ratios
```
What are those "region" and "end region" comments for?

Yosri-Ben-Halima replied:
They kinda make it easy for me to navigate large code files in vs code by partitionning files into subsections. I'll remove them when pushing!

<img width="134" height="586" alt="image" src="https://github.com/user-attachments/assets/5e3229f3-1643-434e-8403-eef94fd0d583" />

konradbachusz replied:
Feel free to remove them. Feel free to use comment blocks like below to make them more visible

```
######################
#### Alpha & Beta ####
######################
```

Yosri-Ben-Halima replied:
Sounds Good

konradbachusz commented on algorithmic_trading_utilities/common/portfolio_ops.py:
```diff
@@ -8,284 +8,295 @@
+        dd = self.drawdown_series()
+        duration = 0
+        max_duration = 0
+        for d in dd:
```
Could we please rewrite this look as a list comprehension, map or apply for performance reasons?

Yosri-Ben-Halima replied:
Yes sure

konradbachusz commented on algorithmic_trading_utilities/common/portfolio_ops.py:
```diff
@@ -8,284 +8,295 @@
+        self.portfolio = original_portfolio
+        self.returns = original_returns
+
+        print()
```
I don't think this empty print is supposed to be there. Please remove it if not needed.

Yosri-Ben-Halima replied:
Oh yes! i will remove that

konradbachusz commented on algorithmic_trading_utilities/common/portfolio_ops.py:
```diff
@@ -8,284 +8,295 @@
+                f"\n{label:{label_width}} {strat_val_fmt:>{value_width}} {bench_val_fmt:>{market_width}}"
+            )
+
+        print("\n" + "=" * span)
```
A nice thing to have here:

Thank you for introducing some new performance metrics, like CVaR that I haven't heard of before 🤯

I want to make this library quite accessible to people with different levels of quant/algo trading understanding. At the bottom of this report could we please maybe include a link to some Benchmark_Interpretation.md file, that can live in this repo, where for each metric we provide a basic explanation of how to interpret different ranges of values?

**Example:**

# alpha:
Definition: <alpha definition and and formula>

## Interpretation
alpha <0: means this...
alpha >0: means that...

.... the same for all of the other metrics

Feel free to use the definitions and explanations from https://www.investopedia.com/financial-term-dictionary-4769738


In the future I want to be able to take all of those performance metrics values and dump them into an LLM.

Yosri-Ben-Halima replied:
Haha thanks! I will add a Benchmark_Interpretation.md file with clear, intuitive, yet concise definitions of each one of these metrics.

konradbachusz commented on algorithmic_trading_utilities/common/portfolio_ops.py:
```diff
@@ -8,284 +8,295 @@
+
+        print("\n" + "=" * span)
+
+    # end region
```
Can we please remove any unused code?

Yosri-Ben-Halima replied:
Yes I will do a code clean-up

konradbachusz commented on tests/test_portfolio_ops.py:
```diff
@@ -1,164 +1,117 @@
+
+    def test_return_distribution_stats(self):
+        stats = self.pm.return_distribution_stats(alpha=0.05)
+        assert "skewness" in stats
```
Could we please keep one assert per test? This makes it easier to debug broken tests in the future.

Yosri-Ben-Halima replied:
Absolutely, I will refactor it asap !

konradbachusz commented:
I'm using this library in other private GitHub repos. Did you confirm that you're able to import it and use your updated functions in portfolio_ops in another repo?

konradbachusz reviewed (commented):
Added a few comments. Thank you for your contribution @Yosri-Ben-Halima!

konradbachusz commented:
Can we please update the README.md based on your recent changes @Yosri-Ben-Halima ? You can use GenAI for that

konradbachusz commented on algorithmic_trading_utilities/common/portfolio_ops.py:
```diff
@@ -8,284 +8,295 @@
+        model = LinearRegression().fit(X.reshape(-1, 1), y.reshape(-1, 1))
+        return {"alpha": model.intercept_[0], "beta": model.coef_[0][0]}
+
+    def rolling_alpha_beta(self, window=252) -> pd.DataFrame:
```
If this would return a series of rolling beta and alpha. Could we please reuse/add any functions from viz_ops.py to plot those when calling get_portfolio_and_benchmark_returns()?

Yosri-Ben-Halima replied:
I’m preparing a full visualization module that will include rolling alpha and beta plots in the upcoming version of viz_ops.py. It is ready for now but i figured it would be better to add it on a separate branch after this one is merged. Would you like me to add it to this branch ?

konradbachusz reviewed (commented):
Added a few more comments

Yosri-Ben-Halima commented:
> Could we please include a CHANGELOG.md in the repo so we can start using semantic versioning for this library and use releases and tags?
> 
> https://www.pyopensci.org/python-package-guide/documentation/repository-files/changelog-file.html

Yes sure i will add that

Yosri-Ben-Halima commented:
> Could we please include a CHANGELOG.md in the repo so we can start using semantic versioning for this library and use releases and tags?

Yosri-Ben-Halima commented:
> Can we please update the README.md based on your recent changes @Yosri-Ben-Halima ? You can use GenAI for that

Yes sure, will do it

Commit:
Add CHANGELOG.md

Commit:
Update .gitignore

Commit:
Add Benchmark_Interpretation.md file with brief intuitive definitions of each metric

Commit:
Add Google-style docstrings and clean code

Commit:
Optimize drawdown duration computation by avoiding the for loop and vectorizing it

Commit:
Breaking down tests into 1 assert per unit test

Commit:
Update class to handle Series instead of arrays and in case benchmark is not provided it defaults to SP500 close prices series as benchmark.

Commit:
Update unit test class to fit last performance evaluation class changes (input type np.ndarrays to pd.Series)

Commit:
Enhance fetching SP500 historic close prices annotations and adding end_date argument

Commit:
Merge branch 'main' of https://github.com/konradbachusz/algorithmic-trading-utilities into performance-metrics-module

Commit:
Update README.md

Commit:
Update CHANGELOG.md

Commit:
feat: add PerformanceViz class for portfolio visualization

- Introduced `PerformanceViz` to plot portfolio equity, drawdowns, return distributions, and rolling metrics.
- Supports benchmark comparison when provided.
- Integrated with `PerformanceMetrics` for seamless metric calculation.
- Includes methods to create all plots at once.

Commit:
Update CHANGELOG.md

Commit:
Add unit tests for PerformanceViz

konradbachusz commented on algorithmic_trading_utilities/common/Benchmark_Interpretation.md:
```diff
@@ -0,0 +1,114 @@
+**Definition:** Reward per unit of downside risk only.  
+**Interpretation:**  
+
+- Focuses on bad (downside) volatility, ignoring upside swings. Higher is better.  
```
Could we please get some magnitude interpretation guidelines?

Yosri-Ben-Halima replied:
Done

konradbachusz commented on algorithmic_trading_utilities/common/Benchmark_Interpretation.md:
```diff
@@ -0,0 +1,114 @@
+## Drawdown Duration
+
+**Definition:** Longest period the portfolio stayed below its previous peak.  
+**Interpretation:**  
```
Could we please include an example?

Yosri-Ben-Halima replied:
Done

konradbachusz commented on algorithmic_trading_utilities/common/Benchmark_Interpretation.md:
```diff
@@ -0,0 +1,114 @@
+## Volatility / Std Dev
+
+**Definition:** Standard deviation of daily returns; measures total risk.  
+**Interpretation:**  
```
Could we please get some magnitude interpretation guidelines or an intuitive example?

Yosri-Ben-Halima replied:
Done

konradbachusz commented on algorithmic_trading_utilities/common/Benchmark_Interpretation.md:
```diff
@@ -0,0 +1,114 @@
+## Conditional Value at Risk (CVaR)
+
+**Definition:** Average equity value loss beyond the VaR threshold.  
+**Interpretation:**  
```
Could we please get some magnitude interpretation guidelines or an intuitive example?

Yosri-Ben-Halima replied:
Done

Yosri-Ben-Halima replied:
Done

konradbachusz commented on algorithmic_trading_utilities/common/Benchmark_Interpretation.md:
```diff
@@ -0,0 +1,114 @@
+## Kurtosis
+
+**Definition:** Measures “tailedness” of returns distribution.  
+**Interpretation:**  
```
Could we please get some magnitude interpretation guidelines or an intuitive example?

Yosri-Ben-Halima replied:
Done

Yosri-Ben-Halima replied:
Done

konradbachusz commented on algorithmic_trading_utilities/common/Benchmark_Interpretation.md:
```diff
@@ -0,0 +1,114 @@
+**Definition:** Annualized return divided by max drawdown.  
+**Interpretation:**  
+
+- Higher → Better risk-adjusted performance relative to drawdown.  
```
Could we please get some magnitude interpretation guidelines or an intuitive example?

Yosri-Ben-Halima replied:
Done

konradbachusz commented on algorithmic_trading_utilities/common/portfolio_ops.py:
```diff
@@ -1,291 +1,286 @@
-        r
-        for r in [
-            (equity[i] - equity[i - 1]) / equity[i - 1] for i in range(1, len(equity))
+# try:
```
Could we please bring back get_portfolio_history, so I can keep evaluating my Alpaca portfolio?

Yosri-Ben-Halima replied:
Done

konradbachusz commented on algorithmic_trading_utilities/common/viz_ops.py:
```diff
@@ -1,70 +1,265 @@
-    plt.legend()
-    plt.grid(True)
-    plt.show()
+    def plot_equity_curve(
```
Can we please add Google-style docstrings to all of the new functions? You can use GenAI to speed this up

Yosri-Ben-Halima replied:
Done

konradbachusz commented on tests/test_portfolio_ops.py:
```diff
@@ -1,164 +1,136 @@
+        assert isinstance(ab["alpha"], float)
+        assert isinstance(ab["beta"], float)
+
+    def test_calculate_benchmark_metrics_alpha(self):
```
Some of those new unit tests seem to be taking a very long time or are stuck. As a result, the GitHub Action pipeline times out. Can you please fix this? https://github.com/konradbachusz/algorithmic-trading-utilities/actions/runs/17598412812

Yosri-Ben-Halima replied:
The problem with that is that when it is running unit tests upon "build" it's getting stuck trying to import modules that no longer exist i.e. ("plot_time_series" in viz_ops.py) because it uses the old version of unit tests

konradbachusz commented on tests/test_viz_ops.py:
```diff
@@ -1,23 +1,64 @@
+
+class TestPerformanceViz:
+    @pytest.fixture
+    def sample_data(self):
```
Can we please move it to conftest.py so it can be re-used by other tests in the future?

Yosri-Ben-Halima replied:
Done

konradbachusz commented on requirements.txt:
```diff
@@ -13,3 +13,5 @@ alpaca-trade-api==3.2.0
 python-dotenv==1.0.0
 multitasking==0.0.11
 yfinance==0.1.66 #TODO fix this error to make it work with more recent version
+scikit-learn==1.7.1
```
I've been having some issues with those dependencies. What Python version are you using? We should probably Add some Dockerfile to this going forward and run tests locally using a specific Python version

Yosri-Ben-Halima replied:
I'm using Python 3.10.4. Yes I agree that would mitigate version mismatch problems

Commit:
Add Google-Style Docstrings & Enhance Equity with Drawdown Plot

Yosri-Ben-Halima commented on algorithmic_trading_utilities/common/Benchmark_Interpretation.md:
```diff
@@ -0,0 +1,114 @@
+**Definition:** Annualized return divided by max drawdown.  
+**Interpretation:**  
+
+- Higher → Better risk-adjusted performance relative to drawdown.  
```
Done

Commit:
Add guidelines and examples to enhance interpretability

Commit:
Finish up Google-Style docstrings

Commit:
Insert back get_portfolio_history to track alpaca portfolio

Commit:
Comment out error generating modules

Commit:
Add rolling sharpe method

Commit:
Add vol and returns heatmap

Commit:
Integrate sample_data pytest fixture in conftest

konradbachusz committed:
Merge branch 'main' into performance-metrics-module

konradbachusz committed:
Update requirements.txt

konradbachusz committed:
Update requirements.txt

alpaca-trade-api==3.2.0

Commit:
Update requirements.txt to add new dependencies and fix TODO comment formatting

Commit:
Add setup.py for package configuration and dependency management

Commit:
Add __init__.py to initialize the algorithmic_trading_utilities package

Commit:
Fix indexing in plot_drawdown_series to use iloc for maximum drawdown values

Commit:
Remove redundancy conflict of numpy and matplotlib dependencies from requirements.txt

Commit:
Update requirements.txt to correct requests and selenium versions

Commit:
Refactor requirements.txt to remove version specifications for dependencies

Commit:
Update requirements.txt to specify versions for all dependencies

Commit:
Update requirements.txt to change alpaca-py and alpaca-trade-api to use minimum version specifications

Commit:
Update requirements.txt to change dependency versions to use compatible version specifications

Commit:
Update requirements.txt to set python-dotenv version to a minimum specification

Commit:
Update get_sp500_prices function to set end_date parameter as optional
